contracts/blender/src/contract.rs

```diff
@@ -0,0 +1,199 @@
+                )
+                .call_flags(CallFlags::default().set_allow_reentry(true))
+                .returns::<core::result::Result<(), PSP22Error>>()
+                .fire()
+                .map_err(|e| BlenderError::InkEnv(format!("{:?}", e)))?
+                .map_err(BlenderError::Psp22)
+        }
+
+        /// Serialize with `ark-serialize::CanonicalSerialize`.
+        pub fn serialize<T: CanonicalSerialize + ?Sized>(t: &T) -> Vec<u8> {
+            let mut bytes = vec![0; t.serialized_size()];
+            t.serialize(&mut bytes[..]).expect("Failed to serialize");
+            bytes.to_vec()
+        }
+
+        /// Call `pallet_snarcos::verify` for the `deposit` relation with `(token_id, value, note)`
+        /// as public input.
+        fn verify_deposit(
+            &self,
+            token_id: TokenId,
+            value: TokenAmount,
+            note: Note,
+            proof: Vec<u8>,
+        ) -> Result<()> {
+            // For now we assume naive input encoding (from typed arguments).
+            let serialized_input = [
+                Self::serialize(&token_id),
+                Self::serialize(&value),
+                Self::serialize(note.as_ref()),
+            ]
+            .concat();
+
+            self.env()
+                .extension()
+                .verify(DEPOSIT_VK_IDENTIFIER, proof, serialized_input, SYSTEM)
+                .map_err(BlenderError::ChainExtension)
+        }
+
+        /// Register a verifying key for one of the `Relation`.
+        ///
+        /// For blendermaster use only.
+        #[ink(message, selector = 8)]
+        pub fn register_vk(&mut self, relation: Relation, vk: Vec<u8>) -> Result<()> {
+            self.ensure_mr_blendermaster()?;
+            let identifier = match relation {
+                Relation::Deposit => DEPOSIT_VK_IDENTIFIER,
+                Relation::Withdraw => WITHDRAW_VK_IDENTIFIER,
+            };
+            self.env()
+                .extension()
+                .store_key(identifier, vk)
+                .map_err(BlenderError::ChainExtension)
+        }
+
+        /// Check if there is a token address registered at `token_id`.
+        #[ink(message, selector = 9)]
+        pub fn registered_token_address(&self, token_id: TokenId) -> Option<AccountId> {
+            self.registered_tokens.get(token_id)
+        }
+
+        /// Register a token contract (`token_address`) at `token_id`.
+        ///
+        /// For blendermaster use only.
+        #[ink(message, selector = 10)]
+        pub fn register_new_token(
+            &mut self,
+            token_id: TokenId,
+            token_address: AccountId,
+        ) -> Result<()> {
+            self.ensure_mr_blendermaster()?;
+            self.registered_tokens
+                .contains(token_id)
+                .not()
+                .then(|| self.registered_tokens.insert(token_id, &token_address))
+                .ok_or(BlenderError::TokenIdAlreadyRegistered)
```

Comment:
A slightly shorter alternative is to use [`insert_return_size`](https://docs.rs/ink_storage/3.3.1/ink_storage/struct.Mapping.html#method.insert_return_size)
```rust
self
  .registered_tokens
  .insert_return_size(token_id, &token_address)
  .ok_or(BlenderError::TokenIdAlreadyRegistered)
```

This should also be slightly betteer in terms of performance as it doesn't do additional FFI call for `.not()` (which I assume calls to the Runtime to check if the keys exists in the `Mapping`).

Reply:
Oh right. Definitely 🙈